master/internal/trial.go

```diff
@@ -1169,6 +1195,17 @@ func (t *trial) terminated(ctx *actor.Context) {
 	}
 }
 
+func (t *trial) MarkPendingGracefulTermination(ctx *actor.Context) {
+	t.PendingGracefulTermination = true
+	for w := range t.preemptionWatchers {
+		select {
+		case w <- struct{}{}:
+		default:
+			ctx.Log().Info("skipping sending preemption signal because channel was full")
+		}
```

Comment:
This seems like not the right behavior; why is not telling the trial to shutdown at all the right call?

Reply:
this wouldn't happen in actor because they have technically unbounded inboxes. since this channel is bounded, it is possible we try to send on it and it is full, but since it is buffered and size of 1 then send should never block unless we recently sent another preemption signal that was not yet received and sent on by grpc (which is also non-blocking operation). so basically we could only ever see this if there were back to back preemption signals before the grpc sender thread got to run.

Reply:
Seems like that would be fixed by sending the message and closing the channel, which makes more sense to me anyway.

Reply:
Sending on the channel may block though.

Reply:
This is to may sure we don’t block on this send

Reply:
oh do you mean sending the first right away (preempt or noop) then closing for the second (always preempt)? that works but I do hate encoding a protocol implicity in the msg and close like that. I guess i could leave a really good comment.

Reply:
Oh, no I was forgetting about the first message.

I mean this: make a 2-message buffered queue.  Send the first message.  Later, when you decide to preempt, send the second message, close the channel, and remove it from the list of watchers.

Right now, `MarkPendingGracefulTermination` leaves the channels open but that seems strange to me.

Reply:
Ah makes sense. To me it's not much different than this, since this really should never happen, but i'll swap it to that.

Reply:
it occured to me afterwards that it might be cleaner and clearer to pass the initial response in the Ask response, and only pass one message through the channel, ever.  That way the watcher responds with the initial message outside of the for loop, which would make the code structure reflect the behavior a teeny bit more.

But I don't think it makes a whole lot of difference.

Reply:
Ok, I did this but just kept the channel as a generic indicator of "one or more bools" to be received, rather that a bool in some cases and chan bool in others.